### PR TITLE
Notebook LPN for sinewave tasks

### PR DESCRIPTION
**Pull request by severinbratus**
Notebook & .py files containing a simple torch implementation of LPN for sinewave tasks (borrowing the generator from the MAML repo)

The Encoder is implemented as DeepSet network (see https://arxiv.org/abs/1703.06114) predicting `z_mu` and `z_logvar` which parametrize a Gaussian in latent space.*

Then, the initial guess `z` is sampled from the encoder-predicted Gaussian (referred as the reparametrization trick in the paper).

Latent Optimisation steps are implemented with GD with a step size `alpha` (hyperparameter).

The Decoder accepts a concatenation of the latent and the inputs (latent first), returns output predictions (Note: *not* a distribution, as with the encoder).

The loss is a sum of two terms: the reconstruction loss of the decoder (in our case MSE up to a constant), and the KL divergence of the latent distribution z wrt the standard Gaussian $N(0, I)$, weighed by `beta` (hyperparam).

All networks are batched.

Tricky technical part: In training, all examples (`pairs`) are processed at once in `forward` call (exploiting batching functionality, so # pairs (N) becomes part of the batch size/shape (B, N)). Hiding the training-target outputs is achieved by constructing a "one-left-out" representation of `pairs`, where all-but-one N-1 pairs are duplicated in a new dimension. So shape (B, N, ...) becomes (B, N, N-1, ...)

The loss is a combination of NLL loss (in our case actually MSE), and KL-divergence of the latent distribution wrt standard Gaussian (prior).

---

*Note: I decided that encoding IO-pairs in isolation would be too ineffective: it is hard to estimate a distribution of possible sinewave functions given only one x-y pair, -- doing so given a few more example pairs should be much better. The DeepSet architecture (see formula below) provides a way to encode sets (permutation-invariant collections) with neural nets. The effect of using DeepSet may be studied by ablating the `rho` networks in the Encoder (if `rho` is identity, then the set embedding is just the sum of embeddings of its parts). Also note that `rho` actually consists of three parts in our case: `rho_0` computes a common representation for both `mu` and `logvar`; `rho_1` and `rho_2` predict `mu` and `logvar` respectively, given the common representation. 

For reference, the DeepSet formula:

$$
f(X) = \rho\left( \sum_{x \in X} \phi(x) \right)
$$

---

Please review & help spot mistakes!

### CODE REVIEW

File: src/models/encoder.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import torch
```
Comment (hannohiss) [inline src/models/encoder.py:1]: In general the rho and phi Deep Sets stuff probably works a lot better. However, I don't think why we need it as a first step / baseline.
As we discussed, I will start with a simpler version of the decoder and then we can check if that works along with the latent optimization training.

File: src/models/lpn.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        # z_mu, z_logvar: (B, N, H)
+        z_mu, z_logvar = self.encoder(pairs_one_left_out)
+        ## print(f"{z_mu.shape=}")
+
+        assert z_mu.shape == z_logvar.shape
+        b, n = pairs.size(0), pairs.size(1)
+        assert z_mu.shape == (b, n, self.d_latent), f"{z_mu.shape} != {(b, n, self.d_latent)}"
+
+        z_sample = self.sample_latents(z_mu, z_logvar)
+        kl_loss = self.kl_divergence(z_mu, z_logvar)
```
Comment (hannohiss) [inline src/models/lpn.py:57]: Am I right that the KL-Divergence is needed to ensure that the latent space is approximately gaussian and not degenerate?

**Review by hannohiss — APPROVED** [review]
Looks nice so far! 🚀

File: src/models/decoder.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
+from src.models.utils import ReluNet
+
+
+class Decoder(ReluNet):
+    # TODO add a batchnorm
```
[the remark this answers is not in the document]
Comment (severinbratus) [reply src/models/decoder.py:5]: Yes that would be true. I actually intended to leave that comment on the encoder (to have a mechanism to average the pair representations, before `rho`). Then later I realised that batchnorm might leak some information away, and lead to the network "cheating" -- since in our setting N (total # of pairs) is part of the batch dimensions, if we were to batchnorm, then information about the target pair (the one left out) may leak in

[the remark this answers is not in the document]
Comment (severinbratus) [reply src/models/decoder.py:5]: So, batchnorm may be done in the *encoder*, but carefully

File: src/models/lpn.py (hunk 2 above)
Comment (severinbratus) [reply src/models/lpn.py:57]: Yes I think this matches my intuition. The encoder is biased to keep the latent distribution close to the prior, which is a std Gaussian

File: src/models/decoder.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (severinbratus) [reply src/models/decoder.py:5]: *nvm, the torch.mean takes care of that

**Review by wskal — COMMENTED** [review]
Looks great. I saw the todo to add dropout - this would be good, and implementing some kind of norm or batch norm like they do in the non-sine tasks of MAML is also worth considering.
